From 1754d7e3eee0cee10ee1a7e5de0ca70507ea7ce2 Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Fri, 10 Jan 2025 18:58:10 +0800
Subject: [PATCH 395/652] WIN2030-14758:perf:Improve GPU mmu operation and
 schedule domain

Changelogs:
1.Refactor vb_ioctl_flush_all();
2.Provide void arch_sync_cache_all(phys_addr_t phys, size_t size) API.
3.map the mmu and dev buffer of the GPU driver as uncaced memory, so that there is no need to flush
  cache frequently. The fluency of the desktop can be improved significantly.
4.In eswin_l2_flush64, call sched_setaffinity() to make the current->pid run on the node of
  the memory to be flushed.
5.In img sysconfig.c, if mem size is larger than ES_MEM_THRESH_OF_FLUSH_CACHE_ALL, then use
  arch_sync_cache_all().
6.Add SCHED_CLUSTER config in arch/riscv/Kconfig
7.Select HAVE_SETUP_PER_CPU_AREA and NEED_PER_CPU_PAGE_FIRST_CHUNK
8.Remove the next-level-cache property from D0CACHE and D1CACHE
9.Fix a bug in sysconfig.c: Invalid NULL variable in "static PHYS_HEAP_FUNCTIONS gsPhysHeapFuncs".

Signed-off-by: linmin <linmin@eswincomputing.com>
Change-Id: I04beae946a20aa56e2e821f714215299fd8cc06e
---
 arch/riscv/Kconfig                            |  22 +++-
 .../dts/eswin/eswin-win2030-arch-d2d.dtsi     |  66 +++++------
 .../boot/dts/eswin/eswin-win2030-arch.dtsi    |   1 -
 arch/riscv/configs/eic7702_defconfig          |  10 +-
 arch/riscv/include/asm/dma-noncoherent.h      |  40 +++++++
 arch/riscv/include/asm/pgtable.h              |   2 +
 arch/riscv/mm/dma-noncoherent.c               |  47 ++++++++
 .../services/server/common/mmu_common.c       |   4 +-
 .../services/server/common/physmem.c          |   4 +-
 .../services/server/env/linux/osfunc.c        |  36 ++++++
 .../services/server/include/osfunc.h          |   2 +
 .../services/system/eswin_cpu/sysconfig.c     |  27 ++++-
 drivers/memory/eswin/es_mmz_vb/mmz_vb.c       |  71 +++---------
 drivers/soc/sifive/sifive_ccache.c            | 108 ++++++++++++++----
 include/linux/dma-map-ops.h                   |   4 +
 include/linux/mem_perf_api.h                  |  24 +++-
 kernel/sched/core.c                           |   1 +
 17 files changed, 338 insertions(+), 131 deletions(-)

diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index 02d015d8850e..7d26860e5616 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -381,6 +381,22 @@ config SCHED_MC
 	  making when dealing with multi-core CPU chips at a cost of slightly
 	  increased overhead in some places. If unsure say N here.
 
+config SCHED_CLUSTER
+	bool "Cluster scheduler support"
+	help
+	  Cluster scheduler support improves the CPU scheduler's decision
+	  making when dealing with machines that have clusters of CPUs.
+	  Cluster usually means a couple of CPUs which are placed closely
+	  by sharing mid-level caches, last-level cache tags or internal
+	  busses.
+
+config SCHED_SMT
+	bool "SMT scheduler support"
+	help
+	  Improves the CPU scheduler's decision making when dealing with
+	  MultiThreading at a cost of slightly increased overhead in some
+	  places. If unsure say N here.
+
 config NR_CPUS
 	int "Maximum number of CPUs (2-512)"
 	depends on SMP
@@ -416,8 +432,10 @@ config NUMA
 	depends on SMP && MMU
 	select ARCH_SUPPORTS_NUMA_BALANCING
 	select GENERIC_ARCH_NUMA
-	select NEED_PER_CPU_EMBED_FIRST_CHUNK
 	select OF_NUMA
+	select HAVE_SETUP_PER_CPU_AREA
+	select NEED_PER_CPU_EMBED_FIRST_CHUNK
+	select NEED_PER_CPU_PAGE_FIRST_CHUNK
 	select USE_PERCPU_NUMA_NODE_ID
 	help
 	  Enable NUMA (Non-Uniform Memory Access) support.
@@ -428,7 +446,7 @@ config NUMA
 config NODES_SHIFT
 	int "Maximum NUMA Nodes (as a power of 2)"
 	range 1 10
-	default "2"
+	default "1"
 	depends on NUMA
 	help
 	  Specify the maximum number of NUMA Nodes available on the target
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-arch-d2d.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-arch-d2d.dtsi
index 2e3228e16ccd..8a6110d6fcb9 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-arch-d2d.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-arch-d2d.dtsi
@@ -24,39 +24,41 @@
 
 &L64 {
 	cpu-map {
-	#if (CHIPLET_AND_DIE & 0x2)
-		cluster0 {
-			core0 {
-				cpu = <&cpu_0>;
+		//socket0 {
+			#if (CHIPLET_AND_DIE & 0x2)
+			cluster0 {
+				core0 {
+					cpu = <&cpu_0>;
+				};
+				core1 {
+					cpu = <&cpu_1>;
+				};
+				core2 {
+					cpu = <&cpu_2>;
+				};
+				core3 {
+					cpu = <&cpu_3>;
+				};
 			};
-			core1 {
-				cpu = <&cpu_1>;
-			};
-			core2 {
-				cpu = <&cpu_2>;
-			};
-			core3 {
-				cpu = <&cpu_3>;
-			};
-		};
-	#endif
-		cluster1 {
-			core0 {
-				cpu = <&cpu_4>;
-			};
-		#ifndef PLATFORM_HAPS
-			core1 {
-				cpu = <&cpu_5>;
-			};
-			core2 {
-				cpu = <&cpu_6>;
-			};
-			core3 {
-				cpu = <&cpu_7>;
-			};
-		#endif
-		};
+			#endif
+		//};
 
+		//socket1 {
+			cluster1 {
+				core0 {
+					cpu = <&cpu_4>;
+				};
+				core1 {
+					cpu = <&cpu_5>;
+				};
+				core2 {
+					cpu = <&cpu_6>;
+				};
+				core3 {
+					cpu = <&cpu_7>;
+				};
+			};
+		//};
 	};
 
 	cpu_4: cpu@4 {
@@ -319,8 +321,6 @@ D1CACHE: cache-controller@22010000 {
 		compatible = "sifive,ccache1", "cache", "sifive,fu740-c000-ccache";
 		interrupt-parent = <&plic1>;
 		interrupts = <1>, <3>, <4>, <2>;
-		//next-level-cache = <&L9 &L10 &L11 &D2MEM>;
-		next-level-cache = <&L9 &L10 &L11>;
 		reg = <0x0 0x22010000 0x0 0x4000 0x0 0x8000000 0x0 0x400000>;
 		reg-names = "control", "sideband";
 		sifive,a-mshr-count = <60>;
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-arch.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-arch.dtsi
index 80c512ffa74a..9ad8702f5547 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-arch.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-arch.dtsi
@@ -331,7 +331,6 @@ D0CACHE: cache-controller@2010000 {
 			compatible = "sifive,ccache1", "cache", "sifive,fu740-c000-ccache";
 			interrupt-parent = <&plic0>;
 			interrupts = <1>, <3>, <4>, <2>;
-			next-level-cache = <&L9 &L10 &L11>;
 			reg = <0x0 0x2010000 0x0 0x4000 0x0 0x8000000 0x0 0x400000>;
 			reg-names = "control", "sideband";
 			sifive,a-mshr-count = <60>;
diff --git a/arch/riscv/configs/eic7702_defconfig b/arch/riscv/configs/eic7702_defconfig
index 176e8c08f175..4298fe7b06c6 100644
--- a/arch/riscv/configs/eic7702_defconfig
+++ b/arch/riscv/configs/eic7702_defconfig
@@ -9,7 +9,6 @@ CONFIG_PREEMPT_VOLUNTARY=y
 CONFIG_IKCONFIG=y
 CONFIG_IKCONFIG_PROC=y
 CONFIG_NUMA_BALANCING=y
-# CONFIG_NUMA_BALANCING_DEFAULT_ENABLED is not set
 CONFIG_CGROUPS=y
 CONFIG_MEMCG=y
 CONFIG_BLK_CGROUP=y
@@ -34,10 +33,10 @@ CONFIG_PERF_EVENTS=y
 CONFIG_SOC_SIFIVE=y
 CONFIG_SOC_VIRT=y
 CONFIG_SMP=y
+CONFIG_SCHED_CLUSTER=y
 CONFIG_NR_CPUS=8
 CONFIG_HOTPLUG_CPU=y
 CONFIG_NUMA=y
-CONFIG_NODES_SHIFT=1
 CONFIG_RISCV_SBI_V01=y
 # CONFIG_RISCV_BOOT_SPINWAIT is not set
 CONFIG_CMDLINE="earlycon=sbi console=ttyS0,115200n8 clk_ignore_unused cma_pernuma=0x20000000 disable_bypass=false firmware_class.path=/lib/firmware/eic7x/"
@@ -443,6 +442,7 @@ CONFIG_HW_RANDOM=y
 CONFIG_HW_RANDOM_VIRTIO=y
 # CONFIG_I2C_COMPAT is not set
 CONFIG_I2C_CHARDEV=y
+CONFIG_I2C_MUX_PCA954x=y
 # CONFIG_I2C_HELPER_AUTO is not set
 CONFIG_I2C_DESIGNWARE_SLAVE=y
 CONFIG_I2C_DESIGNWARE_PLATFORM=y
@@ -466,8 +466,8 @@ CONFIG_SENSORS_PAC1934=y
 CONFIG_WATCHDOG=y
 CONFIG_DW_WATCHDOG=y
 CONFIG_REGULATOR=y
-CONFIG_REGULATOR_ES5340=y
 CONFIG_REGULATOR_TPS549D22=y
+CONFIG_REGULATOR_ES5340=y
 CONFIG_REGULATOR_PCA9450=y
 CONFIG_MEDIA_SUPPORT=y
 CONFIG_MEDIA_SUPPORT_FILTER=y
@@ -815,6 +815,7 @@ CONFIG_CRC_ITU_T=y
 CONFIG_CRC7=y
 CONFIG_DMA_CMA=y
 CONFIG_DMA_NUMA_CMA=y
+CONFIG_CMA_SIZE_MBYTES=512
 CONFIG_PRINTK_TIME=y
 CONFIG_CONSOLE_LOGLEVEL_DEFAULT=15
 CONFIG_CONSOLE_LOGLEVEL_QUIET=15
@@ -828,6 +829,7 @@ CONFIG_DEBUG_MEMORY_INIT=y
 CONFIG_DEBUG_PER_CPU_MAPS=y
 CONFIG_SOFTLOCKUP_DETECTOR=y
 CONFIG_WQ_WATCHDOG=y
+CONFIG_SCHEDSTATS=y
 CONFIG_DEBUG_TIMEKEEPING=y
 CONFIG_DEBUG_RT_MUTEXES=y
 CONFIG_DEBUG_SPINLOCK=y
@@ -842,5 +844,3 @@ CONFIG_RCU_EQS_DEBUG=y
 # CONFIG_FTRACE is not set
 # CONFIG_RUNTIME_TESTING_MENU is not set
 CONFIG_MEMTEST=y
-CONFIG_I2C_MUX=y
-CONFIG_I2C_MUX_PCA954x=y
\ No newline at end of file
diff --git a/arch/riscv/include/asm/dma-noncoherent.h b/arch/riscv/include/asm/dma-noncoherent.h
index 312cfa0858fb..7a9fe8a0e2db 100644
--- a/arch/riscv/include/asm/dma-noncoherent.h
+++ b/arch/riscv/include/asm/dma-noncoherent.h
@@ -21,8 +21,48 @@ struct riscv_nonstd_cache_ops {
 	void (*wback_inv)(phys_addr_t paddr, size_t size);
 };
 
+#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
+typedef enum {
+	FLAT_DDR_MEM = 0,
+	INTERLEAVE_DDR_MEM,
+	SPRAM,
+} eic770x_memory_type_t;
+
+void _do_arch_sync_cache_all(int nid, eic770x_memory_type_t mem_type);
+#endif
+
 extern struct riscv_nonstd_cache_ops noncoherent_cache_ops;
 
 void riscv_noncoherent_register_cache_ops(const struct riscv_nonstd_cache_ops *ops);
 
+
+static inline void arch_get_mem_node_and_type(unsigned long pfn, int *nid, eic770x_memory_type_t *p_mem_type)
+{
+#ifdef CONFIG_NUMA
+	if (unlikely(PFN_IN_SPRAM_DIE0(pfn))) {
+		*nid = 0;
+		*p_mem_type = SPRAM;
+	}
+	else if(unlikely(PFN_IN_SPRAM_DIE1(pfn))) {
+		*nid = 1;
+		*p_mem_type = SPRAM;
+	}
+	else {
+		*nid = pfn_to_nid(pfn);
+		if (CHECK_MEMORY_RANGE_OPFUNC(pfn, MEM, INTPART0) || CHECK_MEMORY_RANGE_OPFUNC(pfn, MEM, INTPART1))
+			*p_mem_type = INTERLEAVE_DDR_MEM;
+		else
+			*p_mem_type = FLAT_DDR_MEM;
+	}
+#else
+	if (unlikely(PFN_IN_SPRAM_DIE0(pfn))) {
+		*nid = 0;
+		*p_mem_type = SPRAM;
+	}
+	else {
+		*nid = 0;
+		*p_mem_type = FLAT_DDR_MEM;
+	}
+#endif
+}
 #endif	/* __ASM_DMA_NONCOHERENT_H */
diff --git a/arch/riscv/include/asm/pgtable.h b/arch/riscv/include/asm/pgtable.h
index 2cfa65e87cd4..90e1fd734f7d 100644
--- a/arch/riscv/include/asm/pgtable.h
+++ b/arch/riscv/include/asm/pgtable.h
@@ -305,6 +305,8 @@ extern struct pt_alloc_ops pt_ops __initdata;
 
 #define CHECK_MEMORY_RANGE_OPFUNC(pfn, range, die)	((pfn & _##range##_PORT_##die##_ADDR_RANGE_BITMASK) == _##range##_PORT_##die##_ADDR_RANGE_VAL)
 
+#define PFN_IN_SPRAM_DIE0(pfn) ((pfn_to_phys(pfn)>=0x59000000) && (pfn_to_phys(pfn)<0x59400000))
+#define PFN_IN_SPRAM_DIE1(pfn) ((pfn_to_phys(pfn)>=0x79000000) && (pfn_to_phys(pfn)<0x79400000))
 /* pha conversion between mem port and sys port or sysllc_port */
 static inline unsigned long convert_pfn_from_mem_to_sys_port(unsigned long pfn)
 {
diff --git a/arch/riscv/mm/dma-noncoherent.c b/arch/riscv/mm/dma-noncoherent.c
index ba27d4765fb7..eec109eaf89f 100644
--- a/arch/riscv/mm/dma-noncoherent.c
+++ b/arch/riscv/mm/dma-noncoherent.c
@@ -12,6 +12,7 @@
 #include <asm/cacheflush.h>
 #include <asm/dma-noncoherent.h>
 #if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
+#include <soc/sifive/sifive_ccache.h>
 #include <linux/iommu.h>
 #endif
 
@@ -253,4 +254,50 @@ void *arch_dma_set_uncached(void *addr, size_t size)
 
 	return NULL;
 }
+#endif
+
+#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
+void _do_arch_sync_cache_all(int nid, eic770x_memory_type_t mem_type)
+{
+	int cpuid, hartid;
+
+#ifdef CONFIG_NUMA
+	if (likely(mem_type == FLAT_DDR_MEM)) {
+		if (nid == 1) {
+			hartid = 4;
+		}
+		else {
+			hartid = 0;
+		}
+
+		cpuid = riscv_hartid_to_cpuid(hartid);
+		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
+	}
+	else if (mem_type == INTERLEAVE_DDR_MEM) {
+		hartid = 0;
+		cpuid = riscv_hartid_to_cpuid(hartid);
+		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
+
+		hartid = 4;
+		cpuid = riscv_hartid_to_cpuid(hartid);
+		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
+	}
+#else
+	if (likely(mem_type == FLAT_DDR_MEM)) {
+		cpuid = smp_processor_id();
+		hartid = cpuid_to_hartid_map(cpuid);
+		ccache_flush_all(&hartid);
+	}
+#endif
+}
+
+void arch_sync_cache_all(phys_addr_t phys, size_t size)
+{
+	int nid;
+	eic770x_memory_type_t mem_type;
+
+	arch_get_mem_node_and_type(phys_to_pfn(phys), &nid, &mem_type);
+	_do_arch_sync_cache_all(nid, mem_type);
+}
+EXPORT_SYMBOL(arch_sync_cache_all);
 #endif
\ No newline at end of file
diff --git a/drivers/gpu/drm/img/img-volcanic/services/server/common/mmu_common.c b/drivers/gpu/drm/img/img-volcanic/services/server/common/mmu_common.c
index e9df944b8546..e94d52a4ffe7 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/server/common/mmu_common.c
+++ b/drivers/gpu/drm/img/img-volcanic/services/server/common/mmu_common.c
@@ -840,8 +840,10 @@ static PVRSRV_ERROR _MMU_PhysMem_RAImportAlloc(RA_PERARENA_HANDLE hArenaHandle,
 	psImport->base = (RA_BASE_T)psMapping->sDevPAddr.uiAddr;
 	psImport->uSize = uiSize;
 
+	#if (MAP_UNCACHED_MEM)
+		arch_sync_dma_for_device(page_to_phys((struct page*)psMapping->sMemHandle.u.pvHandle), PAGE_ALIGN(uiSize), DMA_TO_DEVICE);
+	#endif
 	return PVRSRV_OK;
-
 e1:
 	OSFreeMem(psMapping);
 e0:
diff --git a/drivers/gpu/drm/img/img-volcanic/services/server/common/physmem.c b/drivers/gpu/drm/img/img-volcanic/services/server/common/physmem.c
index ec05b472cc70..36dfff83c0e1 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/server/common/physmem.c
+++ b/drivers/gpu/drm/img/img-volcanic/services/server/common/physmem.c
@@ -175,7 +175,9 @@ PVRSRV_ERROR DevPhysMemAlloc(PVRSRV_DEVICE_NODE	*psDevNode,
 
 		/*Fill the memory with given content */
 		OSDeviceMemSet(pvCpuVAddr, u8Value, ui32MemSize);
-
+		#if (MAP_UNCACHED_MEM)
+			arch_sync_dma_for_device(page_to_phys((struct page*)psMemHandle->u.pvHandle), PAGE_ALIGN(ui32MemSize), DMA_TO_DEVICE);
+		#endif
 		/*Map the page to the CPU VA space */
 		eError = PhysHeapPagesClean(psDevNode->psMMUPhysHeap,
 		                            psMemHandle,
diff --git a/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/osfunc.c b/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/osfunc.c
index 3f9b2effcf7f..55e1c828e634 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/osfunc.c
+++ b/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/osfunc.c
@@ -333,8 +333,35 @@ PVRSRV_ERROR OSPhyContigPagesMap(PHYS_HEAP *psPhysHeap, PG_HANDLE *psMemHandle,
 						void **pvPtr)
 {
 	size_t actualSize = 1 << (PAGE_SHIFT + psMemHandle->uiOrder);
+	#if (!MAP_UNCACHED_MEM)
 	*pvPtr = kmap((struct page*)psMemHandle->u.pvHandle);
+	#else
+	PVRSRV_DEVICE_NODE *psDevNode = PhysHeapDeviceNode(psPhysHeap);
+	struct device *psDev = psDevNode->psDevConfig->pvOSDevice;
+	struct page *page = (struct page*)psMemHandle->u.pvHandle;
+	unsigned long i, num_pages = PAGE_ALIGN(actualSize) / PAGE_SIZE;
+	struct page **pages = kmalloc_node(sizeof(struct page *) * num_pages, GFP_KERNEL, dev_to_node(psDev));
+	struct page **tmp = pages;
+	pgprot_t prot = PAGE_KERNEL;
+	void *vaddr;
+
+	if (page_to_nid(page) != dev_to_node(psDev)) {
+		pr_warn("%s: mem(phys 0x%llx size 0x%lx) and dev are not on the same die\n", __func__,
+			page_to_phys(page) ,uiSize);
+	}
+
+	for (i = 0; i < num_pages; ++i)
+		*tmp++ = page++;
 
+	prot = pgprot_dmacoherent(PAGE_KERNEL);
+	vaddr = vmap(pages, num_pages, VM_MAP, prot);
+	kfree(pages);
+
+	if (!vaddr)
+		return PVRSRV_ERROR_OUT_OF_MEMORY;
+
+	*pvPtr = vaddr;
+	#endif
 	PVR_UNREFERENCED_PARAMETER(psDevPAddr);
 
 	PVR_UNREFERENCED_PARAMETER(actualSize); /* If we don't take an #ifdef path */
@@ -380,7 +407,11 @@ void OSPhyContigPagesUnmap(PHYS_HEAP *psPhysHeap, PG_HANDLE *psMemHandle, void *
 	PVR_UNREFERENCED_PARAMETER(psPhysHeap);
 	PVR_UNREFERENCED_PARAMETER(pvPtr);
 
+	#if (!MAP_UNCACHED_MEM)
 	kunmap((struct page*) psMemHandle->u.pvHandle);
+	#else
+	vunmap(pvPtr);
+	#endif
 }
 
 PVRSRV_ERROR OSPhyContigPagesClean(PHYS_HEAP *psPhysHeap,
@@ -388,6 +419,7 @@ PVRSRV_ERROR OSPhyContigPagesClean(PHYS_HEAP *psPhysHeap,
                                    IMG_UINT32 uiOffset,
                                    IMG_UINT32 uiLength)
 {
+	#if (!MAP_UNCACHED_MEM)
 	PVRSRV_DEVICE_NODE *psDevNode = PhysHeapDeviceNode(psPhysHeap);
 	PVRSRV_ERROR eError = PVRSRV_OK;
 	struct page* psPage = (struct page*) psMemHandle->u.pvHandle;
@@ -428,6 +460,10 @@ PVRSRV_ERROR OSPhyContigPagesClean(PHYS_HEAP *psPhysHeap,
 	kunmap(psPage);
 
 	return eError;
+	#else
+	return PVRSRV_OK;
+	#endif
+
 }
 
 #if defined(__GNUC__)
diff --git a/drivers/gpu/drm/img/img-volcanic/services/server/include/osfunc.h b/drivers/gpu/drm/img/img-volcanic/services/server/include/osfunc.h
index 67ed3d24895b..2e051fcda35c 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/server/include/osfunc.h
+++ b/drivers/gpu/drm/img/img-volcanic/services/server/include/osfunc.h
@@ -92,6 +92,8 @@ CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #include "pmr.h"
 #endif
 
+#include <linux/dma-map-ops.h>
+#define MAP_UNCACHED_MEM 1
 /******************************************************************************
  * Static defines
  *****************************************************************************/
diff --git a/drivers/gpu/drm/img/img-volcanic/services/system/eswin_cpu/sysconfig.c b/drivers/gpu/drm/img/img-volcanic/services/system/eswin_cpu/sysconfig.c
index d3dec65f39c9..9102b4ec61ae 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/system/eswin_cpu/sysconfig.c
+++ b/drivers/gpu/drm/img/img-volcanic/services/system/eswin_cpu/sysconfig.c
@@ -42,10 +42,12 @@
 #if defined(SUPPORT_ION)
 #include "ion_support.h"
 #endif
+#include <asm/dma-noncoherent.h>
 
 #if defined(SUPPORT_VALIDATION) && defined(PDUMP)
 #include "validation_soc.h"
 #endif
+#define ES_MEM_THRESH_OF_FLUSH_CACHE_ALL 0x100000
 
 extern unsigned int _corefreq_div;
 
@@ -55,12 +57,30 @@ IMG_UINT64 *cpu_cache_flush_addr = NULL;
 extern void eswin_l2_flush64(phys_addr_t addr, size_t size);
 #else
 void eswin_l2_flush64(phys_addr_t addr, size_t size) {
-        arch_sync_dma_for_device(addr, size, DMA_TO_DEVICE);
+#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
+	int nid;
+	int cpuid;
+	eic770x_memory_type_t mem_type;
+
+	cpuid = smp_processor_id();
+	arch_get_mem_node_and_type(phys_to_pfn(addr), &nid, &mem_type);
+	if (likely(mem_type == FLAT_DDR_MEM)) {
+		if (cpu_to_node(cpuid) != nid) {
+			pr_debug("%s, pid %d addr 0x%llx, size 0x%lx mem_nid %d cur_cpu %d \n",
+				__func__, task_pid_nr(current), addr, size, nid, smp_processor_id());
+			sched_setaffinity(current->pid, cpumask_of_node(nid));
+		}
+	}
+#endif
+	if (unlikely(size > ES_MEM_THRESH_OF_FLUSH_CACHE_ALL))
+		arch_sync_cache_all(addr, size);
+	else
+		arch_sync_dma_for_device(addr, size, DMA_TO_DEVICE);
 };
 #endif
 void riscv_invalidate_addr(phys_addr_t addr, size_t size,IMG_BOOL virtual) {
 
- printk(KERN_ALERT "eswin:%s not implement now start 0x%x size=0x%x\n",__func__,addr,size);	
+ printk(KERN_ALERT "eswin:%s not implement now start 0x%llx size=0x%lx\n",__func__,addr,size);
 }
 void riscv_flush_addr(IMG_UINT64 cpuaddr,IMG_UINT64 bytes_size, IMG_BOOL virtual)
 {
@@ -180,7 +200,7 @@ static PHYS_HEAP_FUNCTIONS gsPhysHeapFuncs =
 	/* pfnDevPAddrToCpuPAddr */
 	UMAPhysHeapDevPAddrToCpuPAddr,
 	/* pfnGetRegionId */
-	NULL,
+	// NULL,
 };
 
 static PVRSRV_ERROR PhysHeapsCreate(PHYS_HEAP_CONFIG **ppasPhysHeapsOut,
@@ -310,7 +330,6 @@ void riscv_flush_cache_range(IMG_HANDLE hSysData,
 		break;
 	case PVRSRV_CACHE_OP_INVALIDATE:
 		riscv_flush_addr(bStart, (bEnd - bStart), IMG_FALSE);
-		//printk(KERN_ALERT "%s: ---eswin----warning-----PVRSRV_CACHE_OP_INVALIDATE not implement now ------- \n", __func__);
 		break;
 	case PVRSRV_CACHE_OP_CLEAN:
 		riscv_flush_addr(bStart, (bEnd - bStart), IMG_FALSE);
diff --git a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
index b1f201aeaf88..635ee815b7e7 100644
--- a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
+++ b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
@@ -45,7 +45,8 @@
 #include <uapi/linux/es_vb_user.h>
 #include <uapi/linux/mmz_vb.h>
 #include "include/linux/mmz_vb.h"
-#include <soc/sifive/sifive_ccache.h>
+#include <asm/dma-noncoherent.h>
+
 
 MODULE_IMPORT_NS(DMA_BUF);
 
@@ -102,7 +103,7 @@ static int vb_blk_to_pool(struct esVB_BLOCK_TO_POOL_CMD_S *blkToPoolCmd);
 static int vb_get_blk_offset(struct esVB_GET_BLOCKOFFSET_CMD_S *getBlkOffsetCmd);
 static int vb_split_dmabuf(struct esVB_SPLIT_DMABUF_CMD_S *splitDmabufCmd);
 static int vb_get_dmabuf_refcnt(struct esVB_DMABUF_REFCOUNT_CMD_S *getDmabufRefCntCmd);
-static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, bool *interleave_p);
+static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, eic770x_memory_type_t *p_mem_type);
 static int vb_get_dmabuf_size(struct esVB_DMABUF_SIZE_CMD_S *getDmabufSizeCmd);
 static int mmz_vb_pool_exit(void);
 static int mmz_vb_init_memory_region(void);
@@ -855,7 +856,7 @@ static int vb_ioctl_get_dmabuf_refcnt(void __user *user_getDmabufRefCntCmd)
 static int vb_ioctl_retrieve_mem_node(void __user *user_retrieveMemNodeCmd)
 {
 	int ret = 0;
-	bool interleave;
+	eic770x_memory_type_t mem_type;
 	struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd;
 
 	retrieveMemNodeCmd = kzalloc(sizeof(*retrieveMemNodeCmd), GFP_KERNEL);
@@ -867,7 +868,7 @@ static int vb_ioctl_retrieve_mem_node(void __user *user_retrieveMemNodeCmd)
 		goto out_free;
 	}
 
-	ret = vb_retrieve_mem_node(retrieveMemNodeCmd, &interleave);
+	ret = vb_retrieve_mem_node(retrieveMemNodeCmd, &mem_type);
 	if (ret)
 		goto out_free;
 
@@ -1470,9 +1471,7 @@ static int vb_ioctl_uninit_config(void __user *user_cmd)
 
 static int vb_ioctl_flush_all(void __user *user_cmd)
 {
-	int cpuid, hartid;
-#ifdef CONFIG_NUMA
-	bool interleave;
+	eic770x_memory_type_t mem_type;
 	struct esVB_RETRIEVE_MEM_NODE_CMD_S retrieveMemNodeCmd;
 	int ret = 0;
 
@@ -1480,35 +1479,12 @@ static int vb_ioctl_flush_all(void __user *user_cmd)
 		return -EFAULT;
 	}
 	retrieveMemNodeCmd.cpu_vaddr = NULL;
-	ret = vb_retrieve_mem_node(&retrieveMemNodeCmd, &interleave);
+	ret = vb_retrieve_mem_node(&retrieveMemNodeCmd, &mem_type);
 	if (ret)
 		return ret;
 
-	if (likely(interleave == false)) {
-		if (retrieveMemNodeCmd.numa_node == 1) {
-			hartid = 4;
-		}
-		else {
-			hartid = 0;
-		}
+	_do_arch_sync_cache_all(retrieveMemNodeCmd.numa_node, mem_type);
 
-		cpuid = riscv_hartid_to_cpuid(hartid);
-		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
-	}
-	else {
-		hartid = 0;
-		cpuid = riscv_hartid_to_cpuid(hartid);
-		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
-
-		hartid = 4;
-		cpuid = riscv_hartid_to_cpuid(hartid);
-		smp_call_function_single(cpuid, ccache_flush_all, &hartid, true);
-	}
-#else
-	cpuid = smp_processor_id();
-	hartid = cpuid_to_hartid_map(cpuid);
-	ccache_flush_all(&hartid);
-#endif
 	return 0;
 }
 
@@ -2297,17 +2273,13 @@ static int vb_get_dmabuf_refcnt(struct esVB_DMABUF_REFCOUNT_CMD_S *getDmabufRefC
 	return ret;
 }
 
-#define PAGE_IN_SPRAM_DIE0(page) ((page_to_phys(page)>=0x59000000) && (page_to_phys(page)<0x59400000))
-#define PAGE_IN_SPRAM_DIE1(page) ((page_to_phys(page)>=0x79000000) && (page_to_phys(page)<0x79400000))
-static int do_vb_retrive_mem_node(struct dma_buf *dmabuf, int *nid, bool *interleave_p)
+static int do_vb_retrive_mem_node(struct dma_buf *dmabuf, int *p_nid, eic770x_memory_type_t *p_mem_type)
 {
 	int ret = 0;
 	struct dma_buf_attachment *attach;
 	struct sg_table *sgt;
 	struct page *page = NULL;
-#ifdef CONFIG_NUMA
 	unsigned long pfn;
-#endif
 
 	get_dma_buf(dmabuf);
 	attach = dma_buf_attach(dmabuf, mmz_vb_dev);
@@ -2327,33 +2299,20 @@ static int do_vb_retrive_mem_node(struct dma_buf *dmabuf, int *nid, bool *interl
 	}
 
 	page = sg_page(sgt->sgl);
-	if (unlikely(PAGE_IN_SPRAM_DIE0(page))) {
-		*nid = 0;
-	}
-	else if(unlikely(PAGE_IN_SPRAM_DIE1(page))) {
-		*nid = 1;
-	}
-	else
-		*nid = page_to_nid(page);
-
-#ifdef CONFIG_NUMA
 	pfn = page_to_pfn(page);
-	if (CHECK_MEMORY_RANGE_OPFUNC(pfn, MEM, INTPART0) || CHECK_MEMORY_RANGE_OPFUNC(pfn, MEM, INTPART1))
-		*interleave_p = true;
-	else
-		*interleave_p = false;
-#endif
+	arch_get_mem_node_and_type(pfn, p_nid, p_mem_type);
+
 	dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
 	/* detach */
 	dma_buf_detach(dmabuf, attach);
 	/* put dmabuf back */
 	dma_buf_put(dmabuf);
 
-	pr_debug("%s, mem node is %d\n", __func__, *nid);
+	pr_debug("%s, mem node is %d\n", __func__, *p_nid);
 	return ret;
 }
 
-static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, bool *interleave_p)
+static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, eic770x_memory_type_t *p_mem_type)
 {
 	int ret = 0;
 	struct dma_buf *dmabuf;
@@ -2370,7 +2329,7 @@ static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMem
 			return PTR_ERR(dmabuf);
 		}
 
-		ret = do_vb_retrive_mem_node(dmabuf, &retrieveMemNodeCmd->numa_node, interleave_p);
+		ret = do_vb_retrive_mem_node(dmabuf, &retrieveMemNodeCmd->numa_node, p_mem_type);
 		/* put dmabuf back */
 		dma_buf_put(dmabuf);
 	}
@@ -2390,7 +2349,7 @@ static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMem
 			return -EFAULT;
 		}
 		dmabuf = vma->vm_private_data;
-		ret = do_vb_retrive_mem_node(dmabuf, &retrieveMemNodeCmd->numa_node, interleave_p);
+		ret = do_vb_retrive_mem_node(dmabuf, &retrieveMemNodeCmd->numa_node, p_mem_type);
 	}
 
 	return ret;
diff --git a/drivers/soc/sifive/sifive_ccache.c b/drivers/soc/sifive/sifive_ccache.c
index 55e69e613a25..24b0859ea482 100644
--- a/drivers/soc/sifive/sifive_ccache.c
+++ b/drivers/soc/sifive/sifive_ccache.c
@@ -45,7 +45,20 @@ enum {
 	CACHE_NODE_1,
 	SHARE_CACHE_NODE_NUM,
 };
+
+struct ccache_flush_mem_info {
+	int nid;
+	int cpuid;
+	eic770x_memory_type_t mem_type;
+	phys_addr_t paddr;
+	size_t size;
+	unsigned long line;
+};
 static void __iomem *ccache_base[SHARE_CACHE_NODE_NUM];
+DEFINE_PER_CPU(void __iomem* [SHARE_CACHE_NODE_NUM], per_cpu_ccache_base);
+
+DEFINE_PER_CPU(struct ccache_flush_mem_info, per_cpu_ccache_flush_mem_info);
+
 static void __iomem *zero_device_base[SHARE_CACHE_NODE_NUM] = {NULL};
 static int g_irq[SHARE_CACHE_NODE_NUM][SIFIVE_CCACHE_MAX_ECCINTR];
 static struct riscv_cacheinfo_ops ccache_cache_ops;
@@ -102,53 +115,89 @@ static void ccache_way_enable(int node_id)
 	val = FIELD_GET(SIFIVE_CCACHE_CONFIG_WAYS_MASK, cfg);
 	writel(val -1 , ccache_base[node_id] + SIFIVE_CCACHE_WAYENABLE);
 }
+
+#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC7702_SOC)
+static inline void _do_ccache_flush64_range(void *arg)
+{
+	struct ccache_flush_mem_info *p_mem = (struct ccache_flush_mem_info *)arg;
+
+	p_mem->size = p_mem->size + (p_mem->paddr % SIFIVE_CCACHE_FLUSH64_LINE_LEN);
+	p_mem->paddr = ALIGN_DOWN(p_mem->paddr, SIFIVE_CCACHE_FLUSH64_LINE_LEN);
+
+	// pr_debug("%s, mem_nid %d, cpuid %d(smp_processor_id %d)", __func__,
+	// 	 nid, cpudid, smp_processor_id());
+	mb();	/* sync */
+	if (p_mem->mem_type == FLAT_DDR_MEM) {
+		for (p_mem->line = p_mem->paddr; p_mem->line < p_mem->paddr + p_mem->size;
+		p_mem->line += SIFIVE_CCACHE_FLUSH64_LINE_LEN) {
+			writeq_cpu(p_mem->line, (per_cpu(per_cpu_ccache_base, p_mem->cpuid)[p_mem->nid] + SIFIVE_CCACHE_FLUSH64));
+		}
+	}
+	else {
+		for (p_mem->line = p_mem->paddr; p_mem->line < p_mem->paddr + p_mem->size;
+		p_mem->line += SIFIVE_CCACHE_FLUSH64_LINE_LEN) {
+			if((!(!(p_mem->line & 0x40000)))^(!(!(p_mem->line & 0x100)))) {
+				writeq_cpu(p_mem->line, (per_cpu(per_cpu_ccache_base, p_mem->cpuid)[CACHE_NODE_1] + SIFIVE_CCACHE_FLUSH64));
+			}
+			else {
+				writeq_cpu(p_mem->line, (per_cpu(per_cpu_ccache_base, p_mem->cpuid)[CACHE_NODE_0] + SIFIVE_CCACHE_FLUSH64));
+			}
+		}
+	}
+	mb();
+}
+static void ccache_flush64_range(phys_addr_t paddr, size_t size)
+{
+	int nid;
+	int cpuid;
+	eic770x_memory_type_t mem_type;
+	struct ccache_flush_mem_info *p_flush_mem_info;
+
+	cpuid = smp_processor_id();
+	arch_get_mem_node_and_type(phys_to_pfn(paddr), &nid, &mem_type);
+	if (likely(mem_type == FLAT_DDR_MEM) || (mem_type == INTERLEAVE_DDR_MEM)) {
+		p_flush_mem_info = &per_cpu(per_cpu_ccache_flush_mem_info, cpuid);
+		p_flush_mem_info->nid = nid;
+		p_flush_mem_info->cpuid = cpuid;
+		p_flush_mem_info->mem_type = mem_type;
+		p_flush_mem_info->paddr = paddr;
+		p_flush_mem_info->size = size;
+		_do_ccache_flush64_range(p_flush_mem_info);
+	}
+	else {
+		WARN(1, "Sifive ccache: flush64 out of range: %llx(%lx), skip flush\n",
+		     paddr, size);
+		return;
+	}
+}
+#else
 static void ccache_flush64_range(phys_addr_t paddr, size_t size)
 {
 	unsigned long line;
 	size = size + (paddr % SIFIVE_CCACHE_FLUSH64_LINE_LEN);
 	paddr = ALIGN_DOWN(paddr, SIFIVE_CCACHE_FLUSH64_LINE_LEN);
 	mb();	/* sync */
-	#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC7702_SOC)
 	if (paddr >= CONFIG_RISCV_DIE0_CACHED_OFFSET && (paddr + size) <= (CONFIG_RISCV_DIE0_CACHED_OFFSET + CONFIG_RISCV_DIE0_MEM_MAX_SIZE)) {
-	#endif
 		for (line = paddr; line < paddr + size;
 		line += SIFIVE_CCACHE_FLUSH64_LINE_LEN) {
 			writeq_cpu(line, (ccache_base[CACHE_NODE_0] + SIFIVE_CCACHE_FLUSH64));
 		}
 		mb();
-	#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC7702_SOC)
-	}else if (paddr >= CONFIG_RISCV_DIE1_CACHED_OFFSET && (paddr + size) <= (CONFIG_RISCV_DIE1_CACHED_OFFSET + CONFIG_RISCV_DIE1_MEM_MAX_SIZE)) {
-		for (line = paddr; line < paddr + size;
-		line += SIFIVE_CCACHE_FLUSH64_LINE_LEN) {
-			writeq_cpu(line, (ccache_base[CACHE_NODE_1] + SIFIVE_CCACHE_FLUSH64));
-		}
-		mb();
-	}
-	else if (paddr >= CONFIG_RISCV_INTERLEAVE_CACHED_OFFSET && (paddr + size) <= (CONFIG_RISCV_INTERLEAVE_CACHED_OFFSET + CONFIG_RISCV_INTERLEAVE_MEM_MAX_SIZE)){
-		for (line = paddr; line < paddr + size;
-		line += SIFIVE_CCACHE_FLUSH64_LINE_LEN) {
-			if((!(!(line & 0x40000)))^(!(!(line & 0x100)))) {
-				writeq_cpu(line, (ccache_base[CACHE_NODE_1] + SIFIVE_CCACHE_FLUSH64));
-			}
-			else {
-				writeq_cpu(line, ccache_base[CACHE_NODE_0] + SIFIVE_CCACHE_FLUSH64);
-			}
-		}
-		mb();
 	}
 	else {
 		WARN(1, "Sifive ccache: flush64 out of range: %llx(%lx), skip flush\n",
 		     paddr, size);
 		return;
 	}
-	#endif
 }
+#endif
+
 static const struct riscv_nonstd_cache_ops ccache_cmo_ops __initdata = {
 	.wback = &ccache_flush64_range,
 	.inv = &ccache_flush64_range,
 	.wback_inv = &ccache_flush64_range,
 };
-#endif
+#endif //IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
 static const struct of_device_id sifive_ccache_ids[] = {
 	{ .compatible = "sifive,fu540-c000-ccache" },
 	{ .compatible = "sifive,fu740-c000-ccache" },
@@ -354,6 +403,18 @@ void ccache_flush_all(void *arg)
 }
 #endif
 
+static void ccache_per_cpu_data_init(int nid)
+{
+	int i;
+
+	for_each_online_cpu(i) {
+		per_cpu(per_cpu_ccache_base, i)[nid] = ccache_base[nid];
+		pr_debug("---%s, cpu %d, per_cpu_ccache_base[%d] addr 0x%llx\n",
+			__func__, i, nid,
+			(u64)__pa(&per_cpu(per_cpu_ccache_base, i)[nid]));
+	}
+}
+
 static int node_index = 0;
 static int __init sifive_ccache_init(void)
 {
@@ -413,6 +474,7 @@ static int __init sifive_ccache_init(void)
 	#ifdef CONFIG_DEBUG_FS
 		setup_sifive_debug();
 	#endif
+		ccache_per_cpu_data_init(node_index);
 	}
 	return 0;
 err_free_irq:
diff --git a/include/linux/dma-map-ops.h b/include/linux/dma-map-ops.h
index f2fc203fb8a1..fe8778de7829 100644
--- a/include/linux/dma-map-ops.h
+++ b/include/linux/dma-map-ops.h
@@ -366,6 +366,10 @@ static inline pgprot_t dma_pgprot(struct device *dev, pgprot_t prot,
 #ifdef CONFIG_ARCH_HAS_SYNC_DMA_FOR_DEVICE
 void arch_sync_dma_for_device(phys_addr_t paddr, size_t size,
 		enum dma_data_direction dir);
+
+#if IS_ENABLED(CONFIG_ARCH_ESWIN_EIC770X_SOC_FAMILY)
+void arch_sync_cache_all(phys_addr_t phys, size_t size);
+#endif
 #else
 static inline void arch_sync_dma_for_device(phys_addr_t paddr, size_t size,
 		enum dma_data_direction dir)
diff --git a/include/linux/mem_perf_api.h b/include/linux/mem_perf_api.h
index a0357b43d914..110497d4d5af 100644
--- a/include/linux/mem_perf_api.h
+++ b/include/linux/mem_perf_api.h
@@ -1,12 +1,24 @@
-// SPDX-License-Identifier: GPL-2.0-only
+// SPDX-License-Identifier: GPL-2.0
 /*
  * Performance test APIs for ESWIN memory
  *
- * Copyright 2024 Beijing ESWIN Computing Technology Co., Ltd.
- *   Authors:
- *    LinMin<linmin@eswincomputing.com>
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
  *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: Min Lin <linmin@eswincomputing.com>
  */
+
 #ifndef __MEM_PERF_API_H__
 #define __MEM_PERF_API_H__
 
@@ -83,6 +95,8 @@ static inline int memperf_record_cycle_end(struct mem_perf_info *m_perf_i)
 }
 
 #if defined(CONFIG_RISCV)
+// #define RTCCLK 32768
+#define RTCCLK 1000000
 static inline int memperf_print_records(struct list_head *mem_perf_list_head)
 {
     struct mem_perf_info *m_perf_i;
@@ -92,7 +106,7 @@ static inline int memperf_print_records(struct list_head *mem_perf_list_head)
         m_perf_i->cycles_elapased = m_perf_i->cycles_end - m_perf_i->cycles_start;
         total_cycles += m_perf_i->cycles_elapased;
     }
-    PRINT_INFO("Total cycles:%lld, %lld us\n", total_cycles, total_cycles*1000*1000/32768);
+    PRINT_INFO("Total cycles:%lld, %lld us\n", total_cycles, total_cycles*1000*1000/RTCCLK);
     list_for_each_entry(m_perf_i, mem_perf_list_head, node) {
         PRINT_INFO("cycle_elapsed:%lld---%%\%lld.%2lld, %s\n",
             m_perf_i->cycles_elapased, (100*m_perf_i->cycles_elapased)/total_cycles,
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 5f0b82502b2c..2593c05d02c3 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -8430,6 +8430,7 @@ long sched_setaffinity(pid_t pid, const struct cpumask *in_mask)
 	put_task_struct(p);
 	return retval;
 }
+EXPORT_SYMBOL(sched_setaffinity);
 
 static int get_user_cpu_mask(unsigned long __user *user_mask_ptr, unsigned len,
 			     struct cpumask *new_mask)
-- 
2.51.1

